File: proj4a/Insertion-Sort/insertion.pl
:- use_module(library(random)).
:- use_module(library(statistics)).

% Gera uma lista de números aleatórios de tamanho N com elementos entre 0 e MaxValue
generate_random_list(N, MaxValue, List) :-
    length(List, N),
    maplist(random_between(0, MaxValue), List).

% Gera 10 listas aleatórias do mesmo tamanho
generate_random_lists(10, Length, MaxValue, Lists) :-
    length(Lists, 10),
    maplist(generate_random_list(Length, MaxValue), Lists).

% Insertion Sort
insertion_sort(List, Sorted) :-
    insertion_sort_helper(List, [], Sorted).

insertion_sort_helper([], Acc, Acc).
insertion_sort_helper([H|T], Acc, Sorted) :-
    insert(H, Acc, NewAcc),
    insertion_sort_helper(T, NewAcc, Sorted).

insert(X, [], [X]).
insert(X, [Y|T], [X,Y|T]) :-
    X =< Y.
insert(X, [Y|T], [Y|Sorted]) :-
    X > Y,
    insert(X, T, Sorted).

% Ordena e mede o tempo de execução de uma lista
sort_and_time_list(List) :-
    statistics(walltime, [_ | [_]]),
    insertion_sort(List, _Sorted),
    statistics(walltime, [_ | [ExecutionTime]]),
    format('Tempo de execução para lista: ~3f segundos~n', [ExecutionTime / 1000.0]).

% Predicado principal para gerar e ordenar as listas
main :-
    % Definir o tamanho e valor máximo dos elementos das listas
    Length = 1000,
    MaxValue = 100,

    % Gerar 10 listas aleatórias
    generate_random_lists(10, Length, MaxValue, Lists),

    % Ordenar e medir o tempo de execução de cada lista
    maplist(sort_and_time_list, Lists).

File: proj4a/Insertion-Sort/insertion.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void generate_random_list(int list[], int size, int max_value) {
    for (int i = 0; i < size; i++) {
        list[i] = rand() % (max_value + 1);
    }
}

void insertion_sort(int list[], int size) {
    for (int i = 1; i < size; i++) {
        int key = list[i];
        int j = i - 1;
        while (j >= 0 && list[j] > key) {
            list[j + 1] = list[j];
            j = j - 1;
        }
        list[j + 1] = key;
    }
}

int main() {
    srand(time(NULL));
    int size = 10000;

    int lists[10][size];

    // Gerar 10 listas aleatórias
    for (int i = 0; i < 10; i++) {
        generate_random_list(lists[i], size, 100);
    }

    // Ordenar cada lista e medir o tempo de execução
    for (int i = 0; i < 10; i++) {
        clock_t start_time = clock();

        insertion_sort(lists[i], size);

        clock_t end_time = clock();
        double execution_time = ((double) (end_time - start_time)) / CLOCKS_PER_SEC;

        // Exibir o tempo de execução para a lista atual
        printf("Tempo de execução para lista %d: %f segundos\n", i + 1, execution_time);
    }

    return 0;
}
